### PR TITLE
Feature/hatch flower

### PR DESCRIPTION
**Pull request by kc5tjn**
The HatchFlower Module is ready for testing.

### CODE REVIEW

File: robot/src/main/java/us/ilite/robot/modules/HatchFlower.java [hunk 1]
```diff
@@ -1,73 +1,353 @@
...
-    public enum SoleniodStates {
-        // TODO hold solenoid states for each state
-        CAPTURE,
-        RELEASE,
-        PUSH;
+    // Needed for prototype testing
+    // private final Codex<Double, ELogitech310> mController;
+
+    private CommandQueue mCurrentCommandQueue;
+
```
Comment (flybotix) [inline robot/src/main/java/us/ilite/robot/modules/HatchFlower.java:44]: We'll always start the match with either a hatch or cargo.  So the initial state for the hatch will always be 'CAPTURE'.

File: robot/src/main/java/us/ilite/robot/modules/HatchFlower.java [hunk 2]
```diff
@@ -1,73 +1,353 @@
...
+        }
+
+        @Override
+        public void init(double pNow) {
+            solenoidOpenClose.set(this.state.grabber);
+        }
+
+        @Override
+        public boolean update(double pNow) {
+            // always return done, delays handled by the Delay command
```
Comment (flybotix) [inline robot/src/main/java/us/ilite/robot/modules/HatchFlower.java:102]: The 'grab' command is never 'finished'.  It is only 'cancelled'.  This is probably why we'd here a click as the solenoid opened during command init, but didn't stay open afterwards.  Recommend moving the solenoid command to the update() method, and always returning false.

**Review by flybotix — DISMISSED** [review]

**Comment by kc5tjn** [comment]
Jesse,

I made the change to the initial state to CAPTURE for the HatchFlower.  As
for the comment "The 'grab' command is never 'finished'. It is only
'cancelled'......"  I was assuming that the solenoids would hold the last
commanded state, is this not the case?

If the solenoids hold the commanded state, then the idea was that the
commands would set the state on init, then complete on the first update.
The CommandQueue should ensure that the steps execute sequentially.

Thus, when a capture is commanded, the HatchFlower would set the solenoid
states so that the hatch is captured (grabbed), and leave the HatchFlower
in the CAPTURE state, which will persist until the PUSH is commanded.
Since the Push is a transient action, the HatchFlower will only be in the
PUSH state while the push is executing.  The push sequence of commands will
leave the HatchFlower in the RELEASE state, with both solenoids in the
release/reset state.

If I always return false for the grab command then no other commands would
be executed.

Am I missing something?  If you want to talk about it you can reach me at:
jmshapiro@iname.com
571-215-5418

Thanks,
Jeff

**Comment by oconnelc** [comment]
@kc5tjn I'm not sure I understand your comment about the Grab Command. The Grab Command is added to the: 
`ParallelCommand step2= new ParallelCommand(step2List);`

All of the commands in this parallel command will get executed on each update. If the method returns true, then the command is done and will get removed from the queue. Since we always returning false, the command will never be returned but the other commands will be executed.

File: robot/src/main/java/us/ilite/robot/modules/HatchFlower.java [hunk 3]
```diff
@@ -1,73 +1,366 @@
...
+    public HatchFlower() {
+
+        // Needed for prototype testing
+        // The controller has the controller button states
+        // this.mController = pController;
+
+        // TODO Do we need to pass the CAN Addresses in via the constructor?
+        // grabSolenoid = new Solenoid(SystemSettings.kHatchFlowerOpenCloseSolenoidAddress);
+        // pushSolenoid = new Solenoid(SystemSettings.kHatchFlowerExtensionSolenoidAddress);
+        grabSolenoid = new Solenoid(4); //Ball Grabber/Holder (2016 robot)
```
Comment (StephenWelch) [inline robot/src/main/java/us/ilite/robot/modules/HatchFlower.java]: Please re-comment these lines before pushing next time

**Review by DanielTolbert — APPROVED** [review]

**Review by DanielTolbert — DISMISSED** [review]

**Review by StephenWelch — DISMISSED** [review]

**Review by ChangJoseph — APPROVED** [review]

**Review by DanielTolbert — APPROVED** [review]

**Pull request merged**